### PR TITLE
perf: let Compiler worry about and manage file and context timestamps

### LINKED ISSUE #13554: High cost of addFileTimestamps and addContextTimestamps
# Bug report
In very large projects (especially those with child compilers - which our project does due to using `worker-loader`), copying these two maps can be very expensive.

**What is the current behavior?**
`addFileTimestamps ` and `addContextTimestamps` copy huge maps on every recompile. In our project, `compiler.fileTimestamps` has around 300k entries (I believe this is mainly because all ignored files being watched still have a null entry. Additionally, copying this increases the memory cost of these maps. Since these caches are just used internally to `FileSystemInfo`, we should be able to safely share these between `new Compilations`.

**If the current behavior is a bug, please provide the steps to reproduce.**
(1) Run perf CPU trace of webpack compile in very large project
(2) see addFileTimestamps and addContextTimestamps take up large section of trace

Happy to sync more to show trace 1:1, but I dont think I can post our internal trace here

**What is the expected behavior?**
Ideally cost of letting compilation use these timestamps should be low. I am about to open a PR with my proposal, which I think is a very simple solution.

Integrating #13555 to our repo **improves recompile times by about 1-1.5 seconds**

**Other relevant information:**
webpack version:  5.38.1
Node.js version: 16.2.0
Operating System: Windows
Additional tools: -

### PR DESCRIPTION
**Pull request by markjm**
Closes #13554

**What kind of change does this PR introduce?**
Perf improvement, especially in watch mode

**Did you add tests for your changes?**
No, there are no tests in this area, but can try if needed

**Does this PR introduce a breaking change?**
Technically this is a breaking change because we are changing public types. I scraped public github to see if anywhere is using this and couldnt find any, so I think it is a very safe change.

**What needs to be documented once your changes are merged?**
None, types are enough

### CODE REVIEW

**Comment by markjm** [comment]
@sokra thoughts on this one?

**Comment by markjm** [comment]
@sokra I am still applying this as a patch for each new WP version and finding continued perf improvement here. Any concerns with these changes that I can help address?

Thanks

File: lib/FileSystemInfo.js [hunk 1]
```diff
@@ -1063,21 +1064,17 @@ class FileSystemInfo {
...
+	useFileTimestamps(map) {
+		this._fileTimestamps = map;
 		this._cachedDeprecatedFileTimestamps = undefined;
 	}
 
 	/**
 	 * @param {Map<string, FileSystemInfoEntry | "ignore" | null>} map timestamps
 	 * @returns {void}
 	 */
-	addContextTimestamps(map) {
```
Comment (alexander-akait) [inline lib/FileSystemInfo.js:1077]: I'm afraid that it is breaking change

**Comment by markjm** [comment]
Hi - I am wondering what the breaking change is? You mean specifically because of the types change? or is there somewhere using this that I have not been able to find?

**Comment by alexander-akait** [comment]
We rename `addFileTimestamps` and `addContextTimestamps`, but other plugins can use it

**Comment by markjm** [comment]
@alexander-akait  Yea, I do understand that it is possible that the rename/change is breaking this - but there are no uses of this function across all of github so wondering if that is "non-breaking enough". I wonder if you think it might be possible to instead _add_ these functions and keep the `add` variants by adding them to the compiler's timestamp map?

I have continued to maintain this patch in our large repo to much benefit, so I think this is a valuable change if there is a way to incorporate it

**Comment by alexander-akait** [comment]
@markjm We can just add new function and add deprecation for old function (soft or just in changelog) and migrate on the next api internally, so we will don't have breaking change

**Comment by markjm** [comment]
I think it is safe to simply add the old functions back but not use them. Calling this function should work exactly the same in the case a plugin adds some timestamps. They will be added to the compiler timestamp map. I have updated the branch for that

**Review by sokra — APPROVED** [review]

**Comment by sokra** [comment]
Thanks

**Pull request merged**